qiskit/transpiler/synthesis/graysynth.py

```diff
@@ -0,0 +1,304 @@
+                                continue
+                            for ttt in range(len(cnotsp[j])):
+                                cnotsp[j][ttt] ^= cnotsp[qubit][ttt]
+        if ilist == []:
+            continue
+        # See line 18 in pseudo-code of Algorithm 1 in the aforementioned paper
+        # this choice of j maximizes the size of the largest subset (S_0 or S_1)
+        # and the larger a subset, the closer it gets to the ideal in the
+        # Gray code of one CNOT per string.
+        j = ilist[np.argmax([[max(row.count(0), row.count(1)) for row in cnots][k] for k in ilist])]
+        cnots0 = []
+        cnots1 = []
+        for y in list(map(list, zip(*cnots))):
+            if y[j] == 0:
+                cnots0.append(y)
+            elif y[j] == 1:
+                cnots1.append(y)
+        cnots0 = list(map(list, zip(*cnots0)))
+        cnots1 = list(map(list, zip(*cnots1)))
+        if qubit == epsilon:
+            sta.append([cnots1, list(set(ilist).difference([j])), j])
+        else:
+            sta.append([cnots1, list(set(ilist).difference([j])), qubit])
+        sta.append([cnots0, list(set(ilist).difference([j])), qubit])
+    qcir += cnot_synth(state, n_sections)
+    return qcir
+
+
+def cnot_synth(state, n_sections=2):
+    """
+    This function is an implementation of the Patel–Markov–Hayes algorithm
+    for optimal synthesis of linear reversible circuits, as specified by an
+    n x n matrix.
+
+    The algorithm is described in detail in the following paper:
+    "Optimal synthesis of linear reversible circuits."
+    Patel, Ketan N., Igor L. Markov, and John P. Hayes.
+    Quantum Information & Computation 8.3 (2008): 282-294.
+
+    Args:
+        state (list[list] or ndarray): n x n matrix, describing the state
+            of the input circuit
+        n_sections (int): number of sections, used in _lwr_cnot_synth(), in the
+            Patel–Markov–Hayes algorithm. n_sections must be a factor of n_qubits.
+
+    Returns:
+        QuantumCircuit: a CNOT-only circuit implementing the
+            desired linear transformation
+
+    Raises:
+        QiskitError: when variable "state" isn't of type numpy.matrix
+    """
+    if not isinstance(state, (list, np.ndarray)):
+        raise QiskitError('state should be of type list or numpy.ndarray, '
+                          'but was of the type {}'.format(type(state)))
+    state = np.array(state)
+    # Synthesize lower triangular part
+    [state, circuit_l] = _lwr_cnot_synth(state, n_sections)
+    state = np.transpose(state)
+    # Synthesize upper triangular part
+    [state, circuit_u] = _lwr_cnot_synth(state, n_sections)
+    circuit_l.reverse()
+    for i in circuit_u:
+        i.reverse()
+    # Convert the list into a circuit of C-NOT gates
+    circ = QuantumCircuit(state.shape[0])
+    for i in circuit_u + circuit_l:
+        circ.cx(i[0], i[1])
+    return circ
+
+
+def _lwr_cnot_synth(state, n_sections):
```

Comment:
I think the `n_sections` argument should actually be `section_size`, because that is how it is used in the paper, and that is how you seem to be interpreting it in line 259.

Reply:
I agree!